### PR TITLE
Fix bug in main.cpp (penalize_nl=false doesn't work). Supress warning on mingw.

### PR DESCRIPTION
**Pull request by tom7**
Pretty sure this is just a bug, but it's always possible I'm missing something!

### CODE REVIEW

**Comment by DannyDaemonic** [comment]
> this macro is already defined by /usr/lib/gcc/x86_64-w64-mingw32/11/include/c++/x86_64-w64-mingw32/bits/os_defines.h:45.

So many little mingw oddities cropping up. I'm curious though, any idea what's including `os_defines.h`?

File: examples/main/main.cpp [hunk 1]
```diff
@@ -428,7 +430,12 @@ int main(int argc, char ** argv) {
                     last_n_tokens.data() + last_n_tokens.size() - last_n_repeat,
                     last_n_repeat, alpha_frequency, alpha_presence);
                 if (!penalize_nl) {
-                    logits[llama_token_nl()] = nl_logit;
+                  for (size_t idx = 0; idx < candidates_p.size; idx++) {
+                    if (candidates_p.data[idx].id == llama_token_nl()) {
+                      candidates_p.data[idx].logit = nl_logit;
+                      break;
+                    }
+                  }
```
Comment (ggerganov) [inline examples/main/main.cpp:438]: Hmm, the order of `candidates_p` does not seem to be changed.
Why do you think it has?

**Review by ivanstepanovftw — COMMENTED** [review]
I do not see bug here. Order of logits is not changed after penalization

File: examples/main/main.cpp (hunk 1 above)
Comment (tom7) [reply examples/main/main.cpp:438]: It's not that the order of `candiates_p` has changed, although IMO it's not great to rely on the fact that `llama_sample_repetition_penalty` and `llama_sample_frequency_and_presence_penalties` preserve the order (without at least documenting that). The main issue is that on line 418 we've copied the logits into local candidates vector (of llama_token_data_array, not raw floats). So modifying the original logit array from the context does nothing, right?

**Comment by tom7** [comment]
Thanks for the quick look!

> > this macro is already defined by /usr/lib/gcc/x86_64-w64-mingw32/11/include/c++/x86_64-w64-mingw32/bits/os_defines.h:45.
> 
> So many little mingw oddities cropping up. I'm curious though, any idea what's including `os_defines.h`?

Apparently it's via `<string>`:
```
examples/main/main.cpp:27: warning: "NOMINMAX" redefined
   27 | #define NOMINMAX
      |
In file included from /usr/lib/gcc/x86_64-w64-mingw32/11/include/c++/x86_64-w64-mingw32/bits/c++config.h:586,
                 from /usr/lib/gcc/x86_64-w64-mingw32/11/include/c++/string:38,
                 from ./examples/common.h:7,
                 from examples/main/main.cpp:6:
/usr/lib/gcc/x86_64-w64-mingw32/11/include/c++/x86_64-w64-mingw32/bits/os_defines.h:45: note: this is the location of the previous definition
   45 | #define NOMINMAX 1
```

**Comment by martindevans** [comment]
It looks like this PR didn't gain much traction, but I think I just discovered the same bug that this PR fixes.

As I understand it the current situation is:
1. [Get the logits](https://github.com/ggerganov/llama.cpp/blob/04f4b1eb10f3e25750ca3e530265ce2841730e6b/examples/main/main.cpp#L577)
2. [Copy that into candidates vector](https://github.com/ggerganov/llama.cpp/blob/04f4b1eb10f3e25750ca3e530265ce2841730e6b/examples/main/main.cpp#L585)
3. [Save the nl token logit](https://github.com/ggerganov/llama.cpp/blob/04f4b1eb10f3e25750ca3e530265ce2841730e6b/examples/main/main.cpp#L598)
4. Do stuff with `candidates_p`
5. [Overwrite nl token logit](https://github.com/ggerganov/llama.cpp/blob/04f4b1eb10f3e25750ca3e530265ce2841730e6b/examples/main/main.cpp#L607)

However, as far as I can see `logits` is never referenced again after the copy (in step 2). Which means that writing a value to it in step 5 is pointless.

This PR changes it to find the nl token in `candidates_p` and overwrite that value instead.

File: examples/main/main.cpp (hunk 1 above)
Comment (ggerganov) [reply examples/main/main.cpp]: @tom7 Apologise for the delay and not paying more attention to this - I think you are right and we've had this bug for quite some time now.

**Review by ggerganov — APPROVED** [review]
@martindevans Thanks for bringing attention again to this

@tom7 Sorry for the delay of about ~3 months 😄
